# Assignment 1
You should submit the **UniversityNumber.ipynb** file and your final prediction file **UniversityNumber.test.txt** to moodle. Make sure your code does not use your local files and that the results are reproducible. Before submitting, please **1. clean all outputs and 2. run all cells in your notebook and keep all running logs** so that we can check.

# 1 $n$-gram Language Model

In [2]:
!mkdir -p data/lm
!wget -O data/lm/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/train.txt
!wget -O data/lm/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/dev.txt
!wget -O data/lm/test.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/test.txt

The syntax of the command is incorrect.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


## 1.1 Building vocabulary

### Code

In [3]:
with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
    train = dict()
    for line in f:
        line = '<START> ' + line + ' <END>'     # to add <START> and <END> to the start and end of each sentence 
        tmp = line.split()
        for word in tmp:
            if word in train.keys():            # if in dict, increment counter
                train[word] += 1
            else:
                train[word] = 1                 # if not in dict, add new entry

    countUNK = 0
    for key in list(train.keys()):              # find all words with freq < 3, sum their counts
        if train[key] < 3:
            countUNK += train[key]
            del train[key]
    train['<UNK>'] = countUNK
    print("Size of vocabulary = " + str(len(train)))

Size of vocabulary = 26601


### Discussion

Size of vocabulary = 26601

Number of parameters of n-gram models = (size of vocabulary) ^ n = (26601) ^ n
eg unigram: 26601 ^ 1
eg bigram: 26601 ^ 2
eg trigram: 26601 ^ 3

## 1.2 $n$-gram Language Modeling

### Code

In [120]:
import math

unigramDict = dict()
bigramDict = dict()
bigramTrain = dict()

# build unigram
def unigram():
    for key in train.keys():
        unigramDict[key] = float(train[key]) / sum(train.values())

def unigramProb(word):
    if word in unigramDict.keys():
        return unigramDict[word]
    else:
        return unigramDict['<UNK>']

# build bigram
def bigramVocab():
    with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            for i in range(len(tmp) - 1):
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'
                    
                if (tmp[i], tmp[i+1]) in bigramTrain.keys():
                    bigramTrain[(tmp[i], tmp[i+1])] += 1
                else:
                    bigramTrain[(tmp[i], tmp[i+1])] = 1

def bigram():
    for key in bigramTrain.keys():
        bigramDict[key] = float(bigramTrain[key]) / train[key[0]]    # P(w2|w1) = C(w1, w2) / C(w1)
           
unigram()
bigramVocab()
bigram()


In [113]:
def unigramPerplexity(set):
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            # print(tmp)
            logProb = 1
            for word in tmp:
                N += 1
                # logProb *= unigramProb(word)
                logProb += math.log2(unigramProb(word))
                # print(logProb, unigramProb(word), word, line)
            sum += (logProb)
    return 2 ** (-sum / N)

def bigramPerplexity(set):
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            logProb = 1
            for i in range(len(tmp) - 1):
                N += 1
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'
                    
                if (tmp[i], tmp[i+1]) not in bigramDict.keys():
                    prob = 0 / train[key[0]]
                else:
                    prob = bigramDict[(tmp[i], tmp[i+1])]
                # logProb += math.log2(bigramDict[(tmp[i], tmp[i+1])])
                logProb += math.log2(prob)
            sum += (logProb)
    return 2 ** (-sum / N)

print("Unigram train perplexity = " + str(unigramPerplexity("train")))
print("Unigram dev perplexity = " + str(unigramPerplexity("dev")))
print("Bigram train perplexity = " + str(bigramPerplexity("train")))
print("Bigram dev perplexity = " + str(bigramPerplexity("dev")))


Unigram train perplexity = 866.6059171242847
Unigram dev perplexity = 795.3775132279369
Bigram train perplexity = 74.93908934229735


ValueError: math domain error

### Discussion

When calculating the perplexity of unigram on the dev set, even if the words aren't in the training set, they are counted as \<UNK\>, a valid token in the vocabulary.
However, when it comes to the bigram model, some new combinations of words are in the dev set does not exist in the training set. These combinations have a count and probability of zero, making the log likelihood of the model negative infinity which caused error when calculating the perplexity.

## 1.3 Smoothing

### 1.3.1 Add-one (Laplace) smoothing

### Code

In [122]:
bigramSmoothDict = dict()

def bigramSmoothPerplexity(set, k):
    for key in bigramTrain.keys():
        bigramSmoothDict[key] = (float(bigramTrain[key]) + k) / (train[key[0]] + k*26601)    # P(w2|w1) = (C(w1, w2) + 1)/ (C(w1) + |V|)
    
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            logProb = 1
            for i in range(len(tmp) - 1):
                N += 1
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'

                if (tmp[i], tmp[i+1]) not in bigramSmoothDict.keys():
                    prob = (k) / (train[key[0]] + k*26601)
                else:
                    prob = bigramSmoothDict[(tmp[i], tmp[i+1])]
                    
                logProb += math.log2(prob)
            sum += (logProb)
    return 2 ** (-sum / N)

# bigramTrain()
print("Bigram train perplexity = " + str(bigramSmoothPerplexity("train", 1)))
print("Bigram dev perplexity = " + str(bigramSmoothPerplexity("dev", 1)))


Bigram train perplexity = 1405.177331370177
Bigram dev perplexity = 1580.6409074084834


#### Discussion

The perplexity after smoothing is worse than before smoothing in 1.2. This is caused by the nature of add-one smoothing being overly simplistic, shifting too much probability mass from seen to unseen bigrams.

#### 1.3.2 Add-k smoothing

##### Code

In [118]:
print("Bigram train perplexity (k=0.001) = " + str(bigramSmoothPerplexity("train", 0.001)))
print("Bigram train perplexity (k=0.01) = " + str(bigramSmoothPerplexity("train", 0.01)))
print("Bigram train perplexity (k=0.1) = " + str(bigramSmoothPerplexity("train", 0.1)))
print()
print("Bigram dev perplexity (k=0.001) = " + str(bigramSmoothPerplexity("dev", 0.001)))
print("Bigram dev perplexity (k=0.01) = " + str(bigramSmoothPerplexity("dev", 0.01)))
print("Bigram dev perplexity (k=0.1) = " + str(bigramSmoothPerplexity("dev", 0.1)))


Bigram train perplexity (k=0.001) = 92.49453763806177
Bigram train perplexity (k=0.01) = 153.61489113220662
Bigram train perplexity (k=0.1) = 396.8822034902569

Bigram dev perplexity (k=0.001) = 432.7520905911309
Bigram dev perplexity (k=0.01) = 395.3147101966442
Bigram dev perplexity (k=0.1) = 631.6455075022197


##### Discussion

From my result, it seems that the perplexity is lower when k is smaller. When k gets closer to 0, the calculated perplexity is also closer to the perplexity of bigram model without any smoothing, while taking care of unseen bigrams.

### 1.3.3 Linear Interpolation

#### Code

In [176]:
trigramDict = dict()
trigramTrain = dict()
def trigram():
    with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            for i in range(len(tmp) - 2):
                if (tmp[i], tmp[i+1], tmp[i+2]) in trigramTrain.keys():
                    trigramTrain[(tmp[i], tmp[i+1], tmp[i+2])] += 1
                else:
                    trigramTrain[(tmp[i], tmp[i+1], tmp[i+2])] = 1

    for key in trigramTrain.keys():
        trigramDict[key] = float(trigramTrain[key]) / bigramDict[(key[0], key[1])]    # P(w3|w1, w2) = C(w1, w2, w3) / C(w1, w2)

# l1 * P(w3|w1, w2) + l2 * P(w2|w1) + l3 * P(w1)
def linearPerplexity(set, l1, l2, l3):
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            logProb = 1
            for i in range(len(tmp) - 2):
                N += 1
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'
                if tmp[i+2] not in train.keys():
                    tmp[i+2] = '<UNK>'
                    
                # trigram: l1 * P(w3|w1, w2)
                if (tmp[i], tmp[i+1], tmp[i+2]) not in trigramDict.keys():
                    prob = 0
                else:
                    prob = l1 * trigramDict[(tmp[i], tmp[i+1], tmp[i+2])]

                # bigram: l2 * P(w2|w1)
                if (tmp[i], tmp[i+1]) not in bigramDict.keys():
                    prob +=  0
                else:
                    prob += l2 * bigramDict[(tmp[i], tmp[i+1])]
                
                # unigram: l3 * P(w1)
                prob += l3 * unigramProb(tmp[i]) 
                
                logProb += math.log2(prob)

            sum += (logProb)
    return 2 ** (-sum / N)

print("Linear interpolation train perplexity (l1=0.2, l2=0.3, l3=0.5) = " + str(linearPerplexity("train", 0.2, 0.3, 0.5)))
print("Linear interpolation dev perplexity (l1=0.2, l2=0.3, l3=0.5) = " + str(linearPerplexity("dev", 0.2, 0.3, 0.5)))

Linear interpolation train perplexity (l1=0.2, l2=0.3, l3=0.5) = 77.11309086165616
Linear interpolation dev perplexity (l1=0.2, l2=0.3, l3=0.5) = 213.90939133918386


In [177]:
bestPerplexity = float('inf')
bestL1 = 0
bestL2 = 0
bestL3 = 0
for i in range(0, 11):
    for j in range(0, 11):
        for k in range(1, 11):
            if i + j + k == 10:
                l1 = i/10
                l2 = j/10
                l3 = k/10
                
                perplexity = linearPerplexity("dev", l1, l2, l3)
                print("Linear interpolation dev perplexity (l1=" + str(l1) + ", l2=" + str(l2) + ", l3=" + str(l3) + ") = " + str(perplexity))
                if perplexity < bestPerplexity:
                    bestPerplexity = perplexity
                    bestL1 = l1
                    bestL2 = l2
                    bestL3 = l3
                    
print("Best l1, l2, l3 = " + str(bestL1) + ", " + str(bestL2) + ", " + str(bestL3))
print("Best dev perplexity = " + str(bestPerplexity))
      

Linear interpolation dev perplexity (l1=0.0, l2=0.0, l3=1.0) = 1032.1985003767463
Linear interpolation dev perplexity (l1=0.0, l2=0.1, l3=0.9) = 243.76824956673613
Linear interpolation dev perplexity (l1=0.0, l2=0.2, l3=0.8) = 197.8511096838683
Linear interpolation dev perplexity (l1=0.0, l2=0.3, l3=0.7) = 178.50271679085586
Linear interpolation dev perplexity (l1=0.0, l2=0.4, l3=0.6) = 169.57266500096227
Linear interpolation dev perplexity (l1=0.0, l2=0.5, l3=0.5) = 166.97656140743092
Linear interpolation dev perplexity (l1=0.0, l2=0.6, l3=0.4) = 169.79191881349527
Linear interpolation dev perplexity (l1=0.0, l2=0.7, l3=0.3) = 179.04762205739414
Linear interpolation dev perplexity (l1=0.0, l2=0.8, l3=0.2) = 199.0752792987818
Linear interpolation dev perplexity (l1=0.0, l2=0.9, l3=0.1) = 247.19996702558674
Linear interpolation dev perplexity (l1=0.1, l2=0.0, l3=0.9) = 1146.8872226408387
Linear interpolation dev perplexity (l1=0.1, l2=0.1, l3=0.8) = 262.1042599050804
Linear interpolatio

#### Discussion

I tried all combinations of hyperparameters where λ1, λ2, λ3 are upto 1 decimal place.
The best hyperparameter set I got is λ1 = 0.0, λ2 = 0.5,λ3 = 0.5, with a perplexity of 166.97656140743092.
However, I do not rule out the possiblity that other sets of hyperparameters that go beyond 1 decimal place can achieve lower perplexity. If that is the case, there would be infintely more set that would need to be tested for further optimisation.

##### 1.3.4 Optimization

#### Discussion

Bayesian optimization could help find the optimal set of hyperparameters. 

# 2 Word Vectors

In [6]:
def load_embedding_model():
    """ Load GloVe Vectors
        Return:
            wv_from_bin: All 400000 embeddings, each lengh 50
    """
    import gensim.downloader as api
    wv_from_bin = api.load("glove-wiki-gigaword-200")
    print("Loaded vocab size %i" % len(list(wv_from_bin.index_to_key)))
    return wv_from_bin

wv_from_bin = load_embedding_model()

Loaded vocab size 400000


## 2.1 Find most similar word
Use cosine similarity to find the most similar word to each of these words. Report the most similar word and its cosine similarity.

In [7]:
import numpy as np
from numpy.linalg import norm

wordList = ["dog", "whale", "before", "however", "fabricate"]
wordDict = {}
vecKeys = wv_from_bin.index_to_key
for word in wordList:
  sim = -1
  most = ""
  for i in range(len(wv_from_bin)):
    x = wv_from_bin[word]
    cosine = np.dot(x,wv_from_bin[i])/(norm(x)*norm(wv_from_bin[i]))
    curword = vecKeys[i]
    if cosine >= sim and curword != word:
      sim = cosine
      most = curword
  wordDict[word] = [most, sim]
print("Answer format: {word: [most similar word, cosine similarity]}")
print(wordDict)

Answer format: {word: [most similar word, cosine similarity]}
{'dog': ['dogs', 0.81368625], 'whale': ['whales', 0.7918058], 'before': ['after', 0.8931248], 'however': ['although', 0.9336163], 'fabricate': ['fabricating', 0.61840194]}


## 2.2 Finding Analogies
Use vector addition and subtraction to compute target vectors for the analogies below. After computing each target vector, find the top three candidates by cosine similarity. Report the candidates and their similarities to the target vector.

- dog : puppy :: cat : ?:
- speak : speaker :: sing : ?:
- France : French :: England : ?:
- France : wine :: England : ?

In [8]:
# dog-puppy+cat
wordList = [["dog", "puppy", "cat"], ["speak", "speaker", "sing"], ["france", "french", "england"], ["france", "wine", "england"]]

vecKeys = wv_from_bin.index_to_key
for analogy in wordList:
  targetVec = wv_from_bin[analogy[0]] - wv_from_bin[analogy[1]] + wv_from_bin[analogy[2]]
  
  candidateList = []
  similarityList = []
  for i in range(len(wv_from_bin)):
    cosine = np.dot(targetVec,wv_from_bin[i])/(norm(targetVec)*norm(wv_from_bin[i]))
    curword = vecKeys[i]
    
    if len(similarityList) < 3 or (len(similarityList) >= 3 and cosine >= similarityList[2]):
      similarityList.append(cosine) 
      similarityList.sort(reverse=True)
      candidateList.insert(similarityList.index(cosine), curword)
  print(f"{analogy[0]} : {analogy[1]} :: {analogy[2]} : ?:", candidateList[0:3])
  print("corresponding similarities:", similarityList[0:3])
# targetVec = wv_from_bin["dog"] - wv_from_bin["puppy"] + wv_from_bin["cat"]
# print("kitty", np.dot(targetVec,wv_from_bin["kitty"])/(norm(targetVec)*norm(wv_from_bin["kitty"])))
# print("kitten", np.dot(targetVec,wv_from_bin["kitten"])/(norm(targetVec)*norm(wv_from_bin["kitten"])))
# print("dogs", np.dot(targetVec,wv_from_bin["dogs"])/(norm(targetVec)*norm(wv_from_bin["dogs"])))

dog : puppy :: cat : ?: ['cat', 'dog', 'dogs']
corresponding similarities: [0.8246017, 0.745641, 0.6278181]
speak : speaker :: sing : ?: ['sing', 'speak', 'singing']
corresponding similarities: [0.7365669, 0.55140996, 0.548521]
france : french :: england : ?: ['england', 'wales', 'ireland']
corresponding similarities: [0.837254, 0.6498687, 0.6262423]
france : wine :: england : ?: ['england', 'france', 'britain']
corresponding similarities: [0.69291466, 0.6897908, 0.61666137]


# 3 Sentiment analysis

In [9]:
!mkdir -p data/classification
!wget -O data/classification/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/train.txt
!wget -O data/classification/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/dev.txt
!wget -O data/classification/test-blind.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/test-blind.txt

The syntax of the command is incorrect.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


## 3.1 Logistic Regression

### Unigram Features

In [179]:
import numpy as np
from sklearn.linear_model import LogisticRegression

with open('data/classification/train.txt', 'r', encoding="utf-8") as f:
    train = []
    for line in f:
        line = line.split()
        # train.append(line)
        sentiment = line[0]
        words = line[1:]
        print(sentiment, words)

1 ['The', 'Rock', 'is', 'destined', 'to', 'be', 'the', '21st', 'Century', "'s", 'new', '``', 'Conan', "''", 'and', 'that', 'he', "'s", 'going', 'to', 'make', 'a', 'splash', 'even', 'greater', 'than', 'Arnold', 'Schwarzenegger', ',', 'Jean-Claud', 'Van', 'Damme', 'or', 'Steven', 'Segal', '.']
1 ['The', 'gorgeously', 'elaborate', 'continuation', 'of', '``', 'The', 'Lord', 'of', 'the', 'Rings', "''", 'trilogy', 'is', 'so', 'huge', 'that', 'a', 'column', 'of', 'words', 'can', 'not', 'adequately', 'describe', 'co-writer\\/director', 'Peter', 'Jackson', "'s", 'expanded', 'vision', 'of', 'J.R.R.', 'Tolkien', "'s", 'Middle-earth', '.']
1 ['Singer\\/composer', 'Bryan', 'Adams', 'contributes', 'a', 'slew', 'of', 'songs', '--', 'a', 'few', 'potential', 'hits', ',', 'a', 'few', 'more', 'simply', 'intrusive', 'to', 'the', 'story', '--', 'but', 'the', 'whole', 'package', 'certainly', 'captures', 'the', 'intended', ',', 'er', ',', 'spirit', 'of', 'the', 'piece', '.']
1 ['Yet', 'the', 'act', 'is', 'st

### Bigram Features

### GloVe Features

Compare the performance of three types of features on dev set. Report the weighted average precision, recall and F1-score for each feature set.

| Feature | precision | recall | F1-score |
| ----------- | --------- | ------ | -------- |
| unigram     |           |        |          |
| bigram      |           |        |          |
| GloVe       |           |        |          |

## 3.2 Better Feature